---
title: "CEAvignette"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{CEAvignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.width = 10,
  fig.height = 10,
  out.width = "100%"
)
```

```{r setup}
library(CEA)
library(ggplot2)
library(ggsci)
library(ggh4x)
library(ggpubr)
library(tidyverse)
library(lme4)
library(lmerTest)
library(ComplexHeatmap)
library(RColorBrewer)
library(SingleCellExperiment)
```


# Introduction 

CEA is an R package for contrastive eigen-analysis including contrastive PCA (cPCA) and its sparse version (scPCA), contrastive cross-covariance analysis (3CA) and its sparse version (s3CA).


In this vignette, we will use CEA package to reproduce the analysis in the original paper, including

1. Single dataset eta tuning example, generating Fig.2

2. Simulation for contrastive spectral clustering, generating Table 2.

3. Simulation for sparse contrastive spectral clustering, generating Table 3.

4. Simulation for 3CA, generating Fig.3

5. Simulation for s3CA, generating Fig.S1 and Fig.S2

6. Real data application of s3CA generating Fig.4 and Table S3, Fig S3.


# Simulation

## 1. Single dataset eta tuning example, generating Fig.2

```{r, eta tuning example}
eta_range <-  seq(0, 100, 0.2)
Xt <- scale(sim_data_CSC$Xt, scale = FALSE)
Xa <- scale(sim_data_CSC$Xa, scale = FALSE)
meta_t <- sim_data_CSC$covariates_t
eta_tuning_res <-  eta_tuning_general(Xt = Xt, Xa = Xa, eta_range = eta_range)
eta_opt <- eta_tuning_res$eta_opt
p_tuning <- eta_tuning_res$plot+scale_color_cosmic()+labs(title = expression("ARI/Silhouette vs candidate"~eta))


meta_all <-  setNames(data.frame(matrix(ncol = ncol(meta_t)+3, nrow = 0)), c(colnames(meta_t), "cPC1","cPC2","eta"))
for (eta_use in c(eta_opt, seq(from = min(eta_range), to = max(eta_range), length.out = 3))) {
  meta_tmp <- meta_t
  meta_tmp$cPC1 <-  Xt %*% eigen(cov(Xt)-eta_use*cov(Xa))$vectors[,1]
  meta_tmp$cPC2 <-  Xt %*% eigen(cov(Xt)-eta_use*cov(Xa))$vectors[,2]
  meta_tmp$eta <- paste0("eta==",eta_use)
  meta_all <- rbind(meta_all, meta_tmp)
}
meta_all$eta <- factor(meta_all$eta, levels=c("eta==0","eta==1.6","eta==50","eta==100"))
p_eta <- ggplot(meta_all)+
            geom_point(aes(x = cPC1, y = cPC2, color = interesting, shape = nuisance))+
            facet_wrap(vars(eta), labeller = label_parsed, scales= "free")+
            theme_bw()+
            scale_color_aaas()+
            scale_shape_manual(values = c(1, 15))+
            theme(
              strip.background = element_rect(fill = "lightblue", color = "black"),
              strip.text = element_text(color = "black", face = "bold", size = 14),
              text = element_text(size = 14),  # Adjust text size
              axis.title = element_text(size = 16),  # Adjust axis title size
              axis.text = element_text(size = 12),   # Adjust axis text size
              legend.position = "bottom")
print(p_eta)
# ggsave("../inst/SavedFigures/sim_etaTuningEg.pdf", 
#        plot = ggarrange(p_tuning, p_eta, ncol = 2, widths = c(1,1.6), labels = "AUTO"), 
#        width = 12, height = 6, units = "in")
```

## 2. Simulation for contrastive spectral clustering, generating Table 2 
```{r  simulation cPCA, eval=FALSE}
set.seed(111)
mu2_candidates <- c(1, 2)
p1_candidates <- c(5, 10)
eta_range <-  seq(0.1, 50, 0.2)
Sigmat_candidates <- c("Identity", "Toep(0.4)", "Toep(0.5)")
results <- setNames(data.frame(matrix(ncol = 8, nrow = 0)),
                     c("ARI_ctst_interesting", "ARI_ctst_nuisance",
                       "ARI_trt_interesting", "ARI_trt_nuisance","eta",
                        "mu2", "p1", "Sigmat"))
for (mu2 in mu2_candidates) {
  for (p1 in p1_candidates) {
    for(Sigmat in Sigmat_candidates){
        p2 <- 10
        p <- p1+p2
        Sigma_t <- case_when(Sigmat == "Identity" ~ diag(1, p),
                             Sigmat == "Toep(0.4)" ~ toeplitz(0.4^(0:(p-1))),
                             Sigmat == "Toep(0.5)" ~ toeplitz(0.5^(0:(p-1))))
        res <- sim_CSC(n11 = 50, n12 = 25, n21 = 55, n22 = 30, m1tilde = 20, m2tilde = 30,
                       mu1 = 5, mu2 = mu2, theta1 = -5, theta2 = 6, p1 = p1, p2 = p2,
                       sigma_t = 1, sigma_a = 1,
                       Sigma_t = Sigma_t, Sigma_a = NULL,
                       eta = eta_range, repetition = 50, save_plot = FALSE)
        res$mu2 <- mu2
        res$mu <- abs(mu2-5)
        res$p1 <- p1
        res$Sigmat <- Sigmat
        results <- rbind(results, res)
      }
    }
  }
# save(results, file = "../inst/SavedData/sim_cPCA.RData")
```

```{r  simulation cPCA visualization, eval=TRUE}
load(file = "../inst/SavedData/sim_cPCA.RData")
plot_results <- pivot_longer(results, 1:4, names_to = "Type", values_to = "ARI")
plot_results$Analysis <- sapply(plot_results$Type, function(x){strsplit(x,"_")[[1]][2]})
plot_results$Cluster <- sapply(plot_results$Type, function(x){strsplit(x,"_")[[1]][3]})
plot_results$Analysis <- ifelse(plot_results$Analysis=="ctst","Contrastive","Target-only")
plot_results$Cluster <- ifelse(plot_results$Cluster=="interesting","Interesting","Nuisance")
plot_results$eta <- paste("eta==", plot_results$eta, sep = "")
plot_results$mu <- paste("mu==", plot_results$mu, sep = "")
plot_results$p1 <- paste("p[1]==", plot_results$p1, sep = "")
plot_results <- plot_results %>% group_by(mu, p1, Sigmat, Analysis, Cluster) %>%
  summarise(Average_ARI = mean(ARI), SE_ARI = sd(ARI), covarage_high = mean(ARI >= 0.9), covarage_low = mean(ARI <= 0.1))
print(knitr::kable(plot_results, format = "markdown", digits = 3))
p_cPCA <- ggplot(plot_results, aes(x = Cluster,  y = Average_ARI))+
          geom_point(aes(color = Analysis), stat="identity", position ="dodge")+
          geom_errorbar(aes(ymin = Average_ARI - SE_ARI, ymax = Average_ARI + SE_ARI), linewidth = 0.5) +
          facet_nested(Sigmat ~ mu + p1, labeller = label_parsed)+
          labs(x = "", y="Average ARI")+
          theme_bw()+
          theme(
            panel.grid = element_blank(),
            strip.background = element_rect(fill = "lightblue", color = "black"),
            strip.text = element_text(color = "black", face = "bold", size = 14),
            text = element_text(size = 14),  # Adjust text size
            axis.title = element_text(size = 16),  # Adjust axis title size
            axis.text = element_text(size = 12),   # Adjust axis text size
            legend.position = "bottom"
         )+
         scale_color_aaas()
print(p_cPCA)
# ggsave("../inst/SavedFigures/sim_cPCA.pdf", p_cPCA, width = 6, height = 9, units = "in")
```

## 3. Simulation for sparse contrastive spectral clustering, generating Table 3

```{r simulation scPCA, eval = FALSE}
set.seed(111)
mu2_candidates <- c(1, 2)
p3_candidates <- c(100, 300)
eta_range <-  seq(0.1, 10, 0.5)
Sigmat_candidates <- c("Identity", "Toep(0.4)", "Toep(0.5)")
results <- setNames(data.frame(matrix(ncol = 8, nrow = 0)),
                     c("ARI_ctst_interesting", "ARI_ctst_nuisance",
                     "ARI_trt_interesting", "ARI_trt_nuisance","eta",
                      "mu2", "p3", "Sigmat"))
for (mu2 in mu2_candidates) {
  for (p3 in p3_candidates) {
    for(Sigmat in Sigmat_candidates){
      p1 <- 5
      p2 <- 10
      p <- p1+p2+p3
      Sigma_t <- case_when(Sigmat == "Identity" ~ diag(1, p),
                           Sigmat == "Toep(0.4)" ~ toeplitz(0.4^(0:(p-1))),
                           Sigmat == "Toep(0.5)" ~ toeplitz(0.5^(0:(p-1))))
      res <- sim_sCSC(n11 = 50, n12 = 25, n21 = 55, n22 = 30, m1tilde = 20, m2tilde = 30,
                      mu1 = 5, mu2 = mu2, theta1 = -5, theta2 = 6, p1 = p1, p2 = p2, p3 = p3,
                      sigma_t = 1, sigma_a = 1, Sigma_t = Sigma_t, Sigma_a = NULL,
                      eta = eta_range, repetition = 50)
       res$mu2 <- mu2
       res$mu <- abs(mu2-5)
       res$p3 <- p3
       res$Sigmat <- Sigmat
       results <- rbind(results, res)
     }
   }
}
# save(results, file = "../inst/SavedData/sim_scPCA.RData")
```

```{r  simlation scPCA visualization, eval=TRUE}
load(file = "../inst/SavedData/sim_scPCA.RData")
plot_results <- pivot_longer(results, 1:4, names_to = "Type", values_to = "ARI")
plot_results$Analysis <- sapply(plot_results$Type, function(x){strsplit(x,"_")[[1]][2]})
plot_results$Cluster <- sapply(plot_results$Type, function(x){strsplit(x,"_")[[1]][3]})
plot_results$Analysis <- ifelse(plot_results$Analysis=="ctst","Contrastive","Target-only")
plot_results$Cluster <- ifelse(plot_results$Cluster=="interesting","Interesting","Nuisance")
plot_results$eta <- paste("eta==", plot_results$eta, sep = "")
plot_results$mu <- paste("mu==", plot_results$mu, sep = "")
plot_results$p3 <- paste("p[3]==", plot_results$p3, sep = "")
plot_results <- plot_results %>% group_by(mu, p3, Sigmat, Analysis, Cluster) %>%
  summarise(Average_ARI = mean(ARI), SE_ARI = sd(ARI), covarage_high = mean(ARI >= 0.9), covarage_low = mean(ARI <= 0.1))
print(knitr::kable(plot_results, format = "markdown", digits = 3))
p_scPCA <- ggplot(plot_results, aes(x = Cluster,  y = Average_ARI))+
          geom_point(aes(color = Analysis), stat="identity", position ="dodge")+
          geom_errorbar(aes(ymin = Average_ARI - SE_ARI, ymax = Average_ARI + SE_ARI), linewidth = 0.5) +
          facet_nested(Sigmat ~ mu + p3, labeller = label_parsed)+
          labs(x = "", y="Average ARI")+
          theme_bw()+
          theme(
            panel.grid = element_blank(),
            strip.background = element_rect(fill = "lightblue", color = "black"),
            strip.text = element_text(color = "black", face = "bold", size = 14),
            text = element_text(size = 14),  # Adjust text size
            axis.title = element_text(size = 16),  # Adjust axis title size
            axis.text = element_text(size = 12),   # Adjust axis text size
            legend.position = "bottom"
         )+
         scale_color_aaas()
print(p_scPCA)
# ggsave("../inst/SavedFigures/sim_scPCA.pdf", p_scPCA, width = 6, height = 9, units = "in")
```

## 4. Simulation for 3CA, generating Fig.3

```{r simulation for 3CA, eval=FALSE}
n_candidates <- c(1000, 2000, 3000)
AR_candidates <- c(0.5, 0.7, 0.9)
results <- setNames(data.frame(matrix(ncol = 6, nrow = 0)), 
                    c("est_trt_cca", "est_ctst_cca", "est_ctst_pca", "eta_est", 
                      "SampleSize", "AR"))
for (n in n_candidates) {
  for (AR in AR_candidates){
    res <- sim_3CA(nt = n, na = n, p = 40, q = 40, s = 10, AR = AR,
                   lambda_t = c(6, 4, 2, 0), lambda_a = c(2, 0.2), eta = NULL,
                   lm_sigma = 1, lm_beta = 1.5, repetition = 50)
    rownames(res) <- NULL
    res$SampleSize <- n
    res$AR <- AR
    results <- rbind(results, res)
  }
}
# save(results, file = "../inst/SavedData/sim_3CA.RData")
```

```{r simulation for 3CA visualization}
load("../inst/SavedData/sim_3CA.RData")
plot_results <- pivot_longer(results, 1:3, names_to = "Type", values_to = "Estimation")
plot_results$Analysis <- sub(".*?_", "", plot_results$Type)
plot_results$SampleSize <- paste("n=", plot_results$SampleSize, sep = "")
plot_results$AR <- paste("AR=", plot_results$AR, sep = "")
plot_results$Analysis <- factor(plot_results$Analysis, labels = c("ContrastiveCCA", "ContrastivePCA", "TreatmentCCA"))
## boxplot
p_3CA <- ggplot(data = plot_results)+
            geom_boxplot(aes(x = factor(Analysis, levels=c("ContrastiveCCA", "ContrastivePCA", "TreatmentCCA"), labels=c("3CA", "cPCA", "trtCCA")), y = abs(Estimation), fill = Analysis), show.legend = FALSE)+
            facet_grid(cols = vars(SampleSize), rows = vars(AR))+
            geom_hline(yintercept = 1.5, linetype = "dashed", color="red")+
            scale_y_continuous(breaks = c( 0, 1.5))+
            theme_bw()+
            theme(
              panel.grid = element_blank(),
              strip.background = element_rect(fill = "lightblue", color = "black"),
              strip.text = element_text(color = "black", face = "bold", size = 14),
              text = element_text(size = 14),  # Adjust text size
              axis.title = element_text(size = 16),  # Adjust axis title size
              axis.text = element_text(size = 12),   # Adjust axis text size
              # axis.text.x = element_text(angle = 45, hjust = 1)
            )+
            scale_fill_d3()+
            labs(y = "Estimation of the coefficient in (6.2)",x="Analysis")
print(p_3CA)
# ggsave("../inst/SavedFigures/sim_3CA.pdf", plot = p_3CA, width = 10, height = 10, units = "in")

```


## 5. Simulation for s3CA, generating Fig.S1 and Fig.S2

```{r simulation for s3CA, eval=FALSE}
n_candidates <- c(3000)
AR_candidates <- c(0.5, 0.7, 0.9)
p_candidates <- c(300, 350, 400)
s_candidates <- c(5,10)
results <- setNames(data.frame(matrix(ncol = 7, nrow = 0)), 
                    c("est_trt_cca", "est_ctst_cca", "est_ctst_pca", "eta_est",
                      "Dimension", "AR","Sparsity"))
for (n in n_candidates) {
  for (AR in AR_candidates){
    for (p in p_candidates) {
      for (s in s_candidates) {
        res <- sim_s3CA(nt = n, na = n, p = p, q = p, s = s, 
                        lambda_t = c(6, 4, 2, 0), lambda_a = c(2, 0.2), eta = NULL,
                        AR = AR, lm_sigma = 1, lm_beta = 1.5, repetition = 50)
        rownames(res) <- NULL
        res$Dimension <- p
        res$AR <- AR
        res$Sparsity <- s
        results <- rbind(results, res)
      }
    }
  }
}
# save(results, file = "../inst/SavedData/sim_s3CA.RData")
```

```{r simulation for s3CA visualization}
load("../inst/SavedData/sim_s3CA.RData")
plot_results <- pivot_longer(results, 1:3, names_to = "Type", values_to = "Estimation")
plot_results$Analysis <- sub(".*?_", "", plot_results$Type)
plot_results$Dimension <- paste("p=", plot_results$Dimension, sep = "")
plot_results$AR <- paste("AR=", plot_results$AR, sep = "")
plot_results$Analysis <- factor(plot_results$Analysis, labels = c("ContrastiveCCA", "ContrastivePCA", "TargetCCA"))
plot_results$Sparsity <- paste("# of non-zero elements =", plot_results$Sparsity)

p_s3CA <- ggplot(data = plot_results[plot_results$Sparsity == "# of non-zero elements = 10",])+
      geom_boxplot(aes(x = factor(Analysis, levels=c("ContrastiveCCA", "ContrastivePCA", "TargetCCA"), labels=c("s3CA", "scPCA", "trtsCCA")), y = abs(Estimation), fill = Analysis), show.legend = FALSE)+
      facet_nested(AR~Sparsity+Dimension)+
      geom_hline(yintercept = 1.5, linetype = "dashed", color="red")+
      scale_y_continuous(breaks = c( 0, 1.5))+
      theme_bw()+
      theme(
        panel.grid = element_blank(),
        strip.background = element_rect(fill = "lightblue", color = "black"),
        strip.text = element_text(color = "black", face = "bold", size = 14),
        text = element_text(size = 14),  # Adjust text size
        axis.title = element_text(size = 16),  # Adjust axis title size
        axis.text = element_text(size = 12),   # Adjust axis text size
      )+
      scale_fill_d3()+
      labs(y = "Estimation of the coefficient in (6.2)",x="Analysis")
print(p_s3CA)
# ggsave("../inst/SavedFigures/sim_s3CA_s10.pdf", plot = p_s3CA, width = 10, height = 10, units = "in")
# ggsave("../inst/SavedFigures/sim_s3CA_s5.pdf", plot = p_s3CA, width = 10, height = 10, units = "in")
```




# Real data application of cPCA on Leukemia single-cell data.

```{r Leukemia data cPCA}
data("leukemia")
healthy1 <- t(counts(leukemia$healthy1_sce))
patient035 <- t(counts(leukemia$patient035_sce))
pat035_class <- leukemia$patient035_sce$cell_prov %>% factor

pat035_pca <- prcomp(patient035, center = TRUE, scale. = TRUE)
pat035_pca_df <- data.frame(
  PC1 = pat035_pca$x[, 1],
  PC2 = pat035_pca$x[, 2],
  class = factor(pat035_class, levels = c("pre_trans_035", "post_trans_035"),
                 labels=c("Pre","Post")),
  clusters = kmeans(pat035_pca$x[, 1], centers = 2)$cluster
) 
aricode::ARI(pat035_pca_df$class, pat035_pca_df$clusters)

pat035_pca_p <- pat035_pca_df %>%
  ggplot(aes(x = PC1, y = PC2)) +
  geom_point(alpha = 0.5, size = 1) +
  theme_bw()+
  theme(legend.position = "right")+
  theme(panel.grid = element_blank(),
        text = element_text(size = 14),  # Adjust text size
        axis.title = element_text(size = 16),  # Adjust axis title size
        axis.text = element_text(size = 12),
        legend.position = "bottom")+
  scale_color_d3()

pat035_pca_p_wcs <- pat035_pca_df %>%
  ggplot(aes(x = PC1, y = PC2, shape = class, color = factor(clusters))) +
  geom_point(alpha = 0.5, size = 2) +
  theme_bw()+
  labs(color = "Clusters", shape = "Transplant")+
  theme(legend.position = "right")+
  theme(panel.grid = element_blank(),
        text = element_text(size = 14),  # Adjust text size
        axis.title = element_text(size = 16),  # Adjust axis title size
        axis.text = element_text(size = 12),
        legend.position = "bottom")+
  scale_color_d3()

# ggsave("../inst/SavedFigures/Leukemia_PCA035.pdf", pat035_pca_p, width = 4, height = 4, unit = "in")
# ggsave("../inst/SavedFigures/Leukemia_PCA035_wcs.pdf", pat035_pca_p_wcs, width = 5, height = 5, unit = "in")

# cPCA with healthy 1
Xt <- scale(as.matrix(patient035))
Xa <- scale(as.matrix(healthy1))

# eta_opt <- eta_tuning_general(Xa = Xa, Xt = Xt, eta_range = seq(1,25,1))
eta_opt <- 3 # chosen by code above

set.seed(111)
# for speed, directly use eigen decompositon of contrast matrix, cPCA(Xt=Xt, Xa=Xa, eta=eta_opt) will give same results
pat035_heal1_cpca <- eigen(cor(Xt)-eta_opt*cor(Xa))

pat035_heal1_cpca_df <- data.frame(
  cPC1 = Xt%*%pat035_heal1_cpca$vectors[,1],
  cPC2 = Xt%*%pat035_heal1_cpca$vectors[,2],
  class = factor(pat035_class, levels = c("pre_trans_035", "post_trans_035"),
                 labels=c("Pre","Post"))) 

pat035_heal1_cpca_df$clusters = kmeans(pat035_heal1_cpca_df[, c("cPC1")], centers = 2)$cluster
aricode::ARI(pat035_heal1_cpca_df$class, pat035_heal1_cpca_df$clusters)



pat035_heal1_cpca_p <- pat035_heal1_cpca_df %>%
  ggplot(aes(x = cPC1, y = cPC2, colour = factor(clusters), shape = class)) +
  geom_point(alpha = 0.5, size = 2) +
  theme_bw()+
  labs(color = "Clusters", shape = "Transplant") +
  theme(panel.grid = element_blank(),
        text = element_text(size = 14),  # Adjust text size
        axis.title = element_text(size = 16),  # Adjust axis title size
        axis.text = element_text(size = 12),
        legend.position = "bottom")+
  scale_color_d3()


# ggsave("../inst/SavedFigures/Leukemia_cPCA0351.pdf", pat035_heal1_cpca_p, width = 5, height = 5, unit = "in")
```

```{r Leukemia data loading plot}
loadings_df <- data.frame(
  cPC_loadings = abs(pat035_heal1_cpca$vectors[,1]),
  PC_loadings = abs(pat035_pca$rotation[,1])
)

loadings_df <- loadings_df[order(loadings_df$PC_loadings), ]
loadings_df$item <- factor(seq_along(loadings_df$cPC_loadings))

p_loadings <- ggplot(loadings_df, aes(x = item)) +
    geom_bar(aes(y = cPC_loadings, fill = "Loading on cPC1"), stat = "identity", alpha = 0.5, position = "dodge") +
    geom_bar(aes(y = PC_loadings, fill = "Loading on PC1"), stat = "identity", alpha = 0.5, position = "dodge") +
  theme_bw()+
    labs(y = "Loadings", fill = "", x= "Genes") +
    scale_fill_manual(values = c("Loading on cPC1" = "red", "Loading on PC1" = "blue"))+ 
    theme(axis.text.x = element_blank(),legend.position = "bottom")+
  geom_hline(yintercept = c(0.05), color = "black", linetype ="dashed")
print(p_loadings)
# ggsave("../inst/SavedFigures/Leukemia_Loadings.pdf", p_loadings, width = 5, height = 5, unit = "in")

p_c_loadings <- ggplot(loadings_df, aes(x = item)) +
    geom_bar(aes(y = sort(cPC_loadings)), fill = "red", stat = "identity", alpha = 0.5, position = "dodge") +
    theme_bw()+
    labs(y = "Absolute values of loadings on cPC1", fill = "", x= "Genes") +
    theme(axis.text.x = element_blank(),
          panel.grid = element_blank(),
          text = element_text(size = 14),  # Adjust text size
          axis.title = element_text(size = 16),  # Adjust axis title size
          axis.text = element_text(size = 12),
          legend.position = "bottom")+
  geom_hline(yintercept = c(0.05), color = "black", linetype ="dashed")
print(p_c_loadings)
# ggsave("../inst/SavedFigures/Leukemia_cPCAloadings.pdf", p_c_loadings, width = 5, height = 5, unit = "in")

hist_c_loadings <-
  ggplot(loadings_df)+
  geom_histogram(aes(x=cPC_loadings, y=..count../sum(..count..)),
                 fill = "pink", color = "black", alpha = 0.5) +
  theme_bw()+
  labs(y = "Proportions", x = "Absolute values of loadings on cPC1") +
  theme(panel.grid = element_blank(),
        text = element_text(size = 14),  # Adjust text size
        axis.title = element_text(size = 16),  # Adjust axis title size
        axis.text = element_text(size = 12),
        legend.position = "bottom")
print(hist_c_loadings)
# ggsave("../inst/SavedFigures/Leukemia_cPCAloadings_hist.pdf", hist_c_loadings, width = 5, height = 5, unit = "in")

p_t_loadings <- ggplot(loadings_df, aes(x = item)) +
    geom_bar(aes(y = (PC_loadings)), fill="skyblue", stat = "identity", alpha = 0.5, position = "dodge") +
    theme_bw()+
    labs(y = "Absolute values of loadings on PC1", fill = "", x= "Genes") +
    theme(axis.text.x = element_blank(),legend.position = "bottom")+
  geom_hline(yintercept = c(0.05), color = "black", linetype ="dashed")+
    theme(axis.text.x = element_blank(),
          panel.grid = element_blank(),
          text = element_text(size = 14),  # Adjust text size
          axis.title = element_text(size = 16),  # Adjust axis title size
          axis.text = element_text(size = 12),
          legend.position = "bottom")
print(p_t_loadings)
# ggsave("../inst/SavedFigures/Leukemia_PCAloadings.pdf", p_t_loadings, width = 5, height = 5, unit = "in")

hist_t_loadings <-
  ggplot(loadings_df)+
  geom_histogram(aes(x=PC_loadings, y=..count../sum(..count..)),
                 fill = "skyblue", color = "black", alpha = 0.5) +
  theme_bw()+
  labs(y = "Proportions", x = "Absolute values of loadings on PC1") +
  theme(panel.grid = element_blank(),
        text = element_text(size = 14),  # Adjust text size
        axis.title = element_text(size = 16),  # Adjust axis title size
        axis.text = element_text(size = 12),
        legend.position = "bottom")
print(hist_t_loadings)
# ggsave("../inst/SavedFigures/Leukemia_PCAloadings_hist.pdf", hist_t_loadings, width = 5, height = 5, unit = "in")

p_moti1 <- ggarrange(pat035_pca_p_wcs, p_t_loadings, hist_t_loadings, common.legend = TRUE, legend="bottom", nrow = 1, labels = "AUTO", align = "h")
ggsave("../inst/SavedFigures/Leukemia_moti.pdf", p_moti1, width = 13, height = 5, unit = "in")
```

```{r Leukemia data enrichment analysis }
# GO enrichment analysis
library(clusterProfiler)
library(org.Hs.eg.db)
library(enrichplot)

genes_sel <- abs(pat035_heal1_cpca$vectors[,1]) > 0.05

ensembl_ids <- colnames(Xt)[genes_sel]

go_MF <- enrichGO(gene = ensembl_ids,
                      OrgDb = org.Hs.eg.db,
                      keyType = "ENSEMBL",       # Indicates Ensembl Gene IDs are being used
                      ont = "MF",                # Ontology: BP (Biological Process), MF (Molecular Function), CC (Cellular Component)
                      pAdjustMethod = "BH",      # p-value adjustment method
                      pvalueCutoff = 0.05,       # Adjusted p-value cutoff
                      qvalueCutoff = 0.2,
                      readable = TRUE)
go_CC <- enrichGO(gene = ensembl_ids,
                      OrgDb = org.Hs.eg.db,
                      keyType = "ENSEMBL",       
                      ont = "CC",                
                      pAdjustMethod = "BH",      
                      pvalueCutoff = 0.05,      
                      qvalueCutoff = 0.2,
                      readable = TRUE)
go_BP <- enrichGO(gene = ensembl_ids,
                      OrgDb = org.Hs.eg.db,
                      keyType = "ENSEMBL",       
                      ont = "BP",                
                      pAdjustMethod = "BH",      
                      pvalueCutoff = 0.05,      
                      qvalueCutoff = 0.2,
                      readable = TRUE)

p_go_MF <- dotplot(go_MF, showCategory = 20, x = "Count",title = "Top 20 Enriched Molecular Functions")
p_go_BP <- dotplot(go_BP, showCategory = 20, x = "Count",title = "Top 20 Enriched Biological Processes")
p_go_CC <- dotplot(go_CC, showCategory = 20, x = "Count",title = "Top 20 Enriched Cellular Components")
# ggsave("../inst/SavedFigures/Leukemia_GOMF.pdf", p_go_MF, width = 6, height = 10, unit = "in")
# ggsave("../inst/SavedFigures/Leukemia_GOBP.pdf", p_go_BP, width = 6, height = 10, unit = "in")
# ggsave("../inst/SavedFigures/Leukemia_GOCC.pdf", p_go_CC, width = 6, height = 10, unit = "in")
print(p_go_MF)
print(p_go_BP)
print(p_go_CC)


# DO enrichment analysis
# convert Ensembl IDs to Entrez IDs for compatibility
entrez_ids <- bitr(ensembl_ids, fromType = "ENSEMBL", toType = "ENTREZID", OrgDb = org.Hs.eg.db)

do_enrichment <- DOSE::enrichDO(
  gene          = entrez_ids$ENTREZID,
  ont           = "HDO",  # Specify Disease Ontology
  organism = "hsa",
  pAdjustMethod = "BH",
  pvalueCutoff  = 0.05,
  qvalueCutoff  = 0.2,
  readable      = TRUE
)


p_do <- dotplot(do_enrichment, showCategory = 20, x = "Count",title = "Enriched Disease Ontology Terms")
# ggsave("../inst/SavedFigures/Leukemia_DO.pdf", p_do, width = 6, height = 6, unit = "in")
print(p_do)


```



```{r Leukemia data expression heatmap}
Xt_sig_order <- Xt[order(pat035_heal1_cpca_df$clusters), genes_sel]
annotation_col <- HeatmapAnnotation(Status=pat035_heal1_cpca_df[order(pat035_heal1_cpca_df$clusters), "class"], Status=list(nrow=1), col = list(Status = c("Pre-transplant" = "purple", "Post-transplant" = "yellow")))
annotation_row <- rowAnnotation(
  Loadings = pat035_heal1_cpca$vectors[genes_sel,1], 
  col = list(Loadings = colorRamp2(c(-0.2, 0, 0.2), c("navy", "white", "orange"))),
  show_annotation_name = FALSE, 
  annotation_legend_param = list(direction = "horizontal"))
 # pdf("../inst/SavedFigures/Leukemia_heatmap.pdf", width = 11, height= 6)
heat0351_clusters <- Heatmap(t(Xt_sig_order), 
                             heatmap_legend_param = list(direction = "horizontal"),
                             cluster_columns = FALSE, cluster_rows = TRUE, 
                             col = colorRamp2(c(-4, 0, 4), c("green", "white", "red")),
                             name = "Standardized expression",
                             column_split = ifelse(pat035_heal1_cpca_df[order(pat035_heal1_cpca_df$clusters), c("clusters")] == 1, "Contrastive cluster1", "Contrastive cluster2"),
                             right_annotation = annotation_row,
                             bottom_annotation = annotation_col,
                             show_column_names = FALSE,
                             show_row_names = FALSE)
draw(heat0351_clusters, merge_legend = TRUE, heatmap_legend_side = "bottom", annotation_legend_side = "bottom")
 # dev.off()
```

# Real data application of s3CA on COVID19 multi-omics data.

```{r covid data s3CA}
data(covid)
covariates <- covid$covariates
mtb <- covid$mtb
protein <- covid$protein

# transformation
mtb_log <- log(mtb + 1)
protein_log <- log(protein + 1)

#### 1. Choose correlation or covariance ####
# Use correlation matrix or covariance matrix
cormat <- TRUE
# Use assigned eta and zta
eta_assign <- 1 # or NULL to use eta estimation
#### 2. Choose conditions ####
# Take Covid as conditions 
Xa <- scale(mtb_log[covariates$covid == "Non-Covid", ], scale = cormat)
Ya <- scale(protein_log[covariates$covid == "Non-Covid",], scale = cormat)
Xt <- scale(mtb_log[covariates$covid != "Non-Covid",], scale = cormat)
Yt <- scale(protein_log[covariates$covid != "Non-Covid",], scale = cormat)

p <- ncol(Xa)
q <- ncol(Ya)
na <- nrow(Xa)
nt <- nrow(Xt)
Kmax <- min(p, q, nt-1)
covariates_a <- covariates[rownames(Xa),]
covariates_t <- covariates[rownames(Xt),]


#### 3. s3CA  ####

res_s3CA <- cCCA(Xa = Xa, Ya = Ya, Xt = Xt, Yt = Yt, eta  = eta_assign,
                sparse_ctst = seq(0.2, 1, len=20), sparse_trt = seq(0.2, 1, len=40), sparse_ctrl = seq(0.2, 1, len=40))
CCA_ctst <- res_s3CA$CCA_ctst
CCA_ctrl <- res_s3CA$CCA_ctrl
CCA_trt <- res_s3CA$CCA_trt

K_single <- K_t <- K_a <- 1
cCCX <- Xt%*%as.matrix(CCA_ctst$u[,K_single])
cCCY <- Yt%*%as.matrix(CCA_ctst$v[,K_single])

df_score <- data.frame(cCCX = cCCX, cCCY = cCCY, covariates_t)

df_score_t <- data.frame(CCtX = Xt%*%as.matrix(CCA_trt$u[,K_t]), 
                         CCtY = Yt%*%as.matrix(CCA_trt$v[,K_t]), covariates_t)

df_score_a <- data.frame(CCcX = Xa%*%as.matrix(CCA_ctrl$u[,K_a]), 
                         CCcY = Ya%*%as.matrix(CCA_ctrl$v[,K_a]), covariates_a)


#### 4. Visualization ####
# 4.1. scree
p_scree <- ggplot(data = data.frame(Component = 1:Kmax, Variation = CCA_ctst$d/sum(CCA_ctst$d)))+
  geom_point(aes(x = Component, y = Variation))+
  theme_bw()+ 
  labs(y = "Proportion of explained variation")+
  theme(
        panel.grid = element_blank(),
        text = element_text(size = 14),  # Adjust text size
        axis.title = element_text(size = 16),  # Adjust axis title size
        axis.text = element_text(size = 12)
      )
print(p_scree)
# ggsave("../inst/SavedFigures/COVID_scree.pdf", plot = p_scree, width = 5, height = 4)


# 4.2. covariance discrepancy
K <- 30
Wa <- Xa%*%CCA_ctst$u[,1:K]
Za <- Ya%*%CCA_ctst$v[,1:K]
Wt <- Xt%*%CCA_ctst$u[,1:K]
Zt <- Yt%*%CCA_ctst$v[,1:K]

cor_disc <- data.frame(Component = rep(1:K, 2),
                       Cov = c(diag(cov(Wa, Za)), diag(cov(Wt, Zt))),
                       Type = rep(c("Control", "Treatment"), each = K))
p_disc <- ggplot(data = cor_disc) +
  geom_line(aes(x = Component, y = Cov, color = Type))+
  labs(y = "Cov(Xu, Yv)")+
  theme_bw()+
  theme(
        panel.grid = element_blank(),
        text = element_text(size = 14),  # Adjust text size
        axis.title = element_text(size = 16),  # Adjust axis title size
        axis.text = element_text(size = 12),
        legend.position = "bottom"
      )+
  scale_color_aaas()
# ggsave("../inst/SavedFigures/COVID_disc.pdf", plot = p_disc, width = 6, height = 5)
print(p_disc)


# 4.3. score scatter nuisance removal
nuisance <-  "icu"
ARI_removal <- c(spec_clust(Ya%*%CCA_ctrl$v[,1], group = df_score_a[,nuisance]),
                 spec_clust(Yt%*%CCA_trt$v[,1], group = df_score_t[,nuisance]),
                 spec_clust(Yt%*%CCA_ctst$v[,1], group = df_score[,nuisance]))
ARI_removal <- paste("ARI(ICU) = ", ARI_removal)
names(ARI_removal) <- c("Control","Treatment","Contrastive")

# 4.4. score scatter continuous interested regression
interested_y <- "Platelet_K.uL"
pval_sig <- c(summary(score_reg(x = data.frame(x = df_score[,"cCCY"]),
                                y = df_score[,interested_y]))$coefficients[2,"Pr(>|t|)"],
              summary(score_reg(x = data.frame(x = df_score_t[,"CCtY"]),
                                y = df_score_t[,interested_y]))$coefficients[2,"Pr(>|t|)"],
              summary(score_reg(data.frame(x = df_score_a[,"CCcY"]), 
                                y = df_score_a[,interested_y]))$coefficients[2,"Pr(>|t|)"])
pval_sig <- round(pval_sig, 4)
pval_sig <- paste("P-value = ", pval_sig, sep = "")
names(pval_sig) <- c("Contrastive","Treatment","Control")

p_reg <- ggplot(data = data.frame(df_score[,c("cCCY","icu",interested_y)]),
                aes_string(x = "cCCY",y = interested_y))+
  geom_point(aes_string(color = "icu" ))+ 
  geom_smooth(method = "lm", se = TRUE)+
  geom_text(data = data.frame(a=c(3,3),b=c(700, 600), #c(3.5, 3),
                              label=c(ARI_removal["Contrastive"], pval_sig["Contrastive"])), 
            aes(x = a, y = b, label = label))+
  theme_bw()+
  labs(title = "Contrastive", x = "Yv", color = "")+
  theme(panel.grid = element_blank(),
        text = element_text(size = 14),  # Adjust text size
        axis.title = element_text(size = 16),  # Adjust axis title size
        axis.text = element_text(size = 12),
        legend.position = "right")+
  scale_color_d3()+
  ylim(0,850)
p_reg_t <- ggplot(data = data.frame(df_score_t[, c("CCtY", "icu",interested_y)]),
                aes_string(x = "CCtY",y = interested_y))+
  geom_point(aes_string(color = "icu" ))+  
  geom_smooth(method = "lm", se = TRUE)+
  geom_text(data = data.frame(a=c(-10,-10),b=c(700, 600), #c(3.5, 3),
                              label=c(ARI_removal["Treatment"], pval_sig["Treatment"])), 
            aes(x = a, y = b, label = label))+
  theme_bw()+
  labs(title = "COVID", x = "Yv", color = "")+
  theme(panel.grid = element_blank(),
        text = element_text(size = 14),  # Adjust text size
        axis.title = element_text(size = 16),  # Adjust axis title size
        axis.text = element_text(size = 12),
        legend.position = "right")+
  scale_color_d3()+
  ylim(0,850)
p_reg_a <- ggplot(data = data.frame(df_score_a[, c("CCcY", "icu",interested_y)]),
                aes_string(x = "CCcY",y = interested_y))+
  geom_point(aes_string(color = "icu" ))+ 
  geom_smooth(method = "lm", se = TRUE)+
  geom_text(data = data.frame(a=c(-15,-15),b=c(700, 600), #c(313, 267), #,c(3.5, 3)
                              label=c(ARI_removal["Control"], pval_sig["Control"])), 
            aes(x = a, y = b, label = label))+
  theme_bw()+
  labs(title = "Non COVID", x = "Yv", color = "")+
  theme(panel.grid = element_blank(),
        text = element_text(size = 14),  # Adjust text size
        axis.title = element_text(size = 16),  # Adjust axis title size
        axis.text = element_text(size = 12),
        legend.position = "right")+
  scale_color_d3()+
  ylim(0,850)

p_score <- ggarrange(p_reg_t, p_reg_a, p_reg, common.legend = T, labels = c("A","B","C"),
                     legend = "bottom", nrow=1, align = "hv")
print(p_score)
# ggsave("../inst/SavedFigures/COVID_score.pdf", plot = p_score, width = 12, height = 6)


# 4.5 loading plot
p_loading <- ggplot(data = data.frame(index = 1:ncol(Yt),
                                      Loadings = sort(abs(CCA_ctst$v[,1])),
                                      mark = rep(c("Others","Top 11"), c(ncol(Yt)-11,11))))+
  geom_point(aes(x=index, y=Loadings, color = mark))+
  theme_bw()+
  labs( x = "Proteins", y = "Absolute values of loadings", color = "")+
  theme(panel.grid = element_blank(),
        text = element_text(size = 14),  # Adjust text size
        axis.title = element_text(size = 16),  # Adjust axis title size
        axis.text = element_text(size = 12),
        legend.position = "bottom")+
  scale_color_d3()
print(p_loading)

# ggsave("../inst/SavedFigures/COVID_loadings.pdf", plot = p_loading, width = 5, height = 4)

print(colnames(Yt)[order(abs(CCA_ctst$v[,1]), decreasing = T)][1:11])

# 4.6 Score plot and comparison of loadings between covid and non-covid group for motivation
p_score_t <- ggplot(data = data.frame(df_score_t[, c("CCtX", "CCtY", "icu")]),
                aes_string(x = "CCtX",y = "CCtY", color = "icu"))+
  geom_point()+  
  theme_bw()+
  labs(title = "", x = "First component with respect to metabolite",  
       y = "First component with respect to protein",
       color = "ICU")+
  theme(panel.grid = element_blank(),
        text = element_text(size = 14),  # Adjust text size
        axis.title = element_text(size = 12),  # Adjust axis title size
        axis.text = element_text(size = 12),
        legend.position = "bottom")+
scale_color_aaas()
p_score_a <- ggplot(data = data.frame(df_score_a[, c("CCcX", "CCcY", "icu")]),
                aes_string(x = "CCcX",y = "CCcY", color = "icu"))+
  geom_point()+  
  theme_bw()+
  labs(title = "", x = "First component with respect to metabolite",  
       y = "First component with respect to protein",
       color = "ICU")+
  theme(panel.grid = element_blank(),
        text = element_text(size = 14),  # Adjust text size
        axis.title = element_text(size = 12),  # Adjust axis title size
        axis.text = element_text(size = 12),
        legend.position = "bottom")+
  scale_color_aaas()


# ggsave("../inst/SavedFigures/COVID_covid_scoreXY.pdf", plot = p_score_t, width = 5, height = 5)
# ggsave("../inst/SavedFigures/COVID_noncovid_scoreXY.pdf", plot = p_score_a, width = 5, height = 5)

p_Yloading_ta_comp <- ggplot(data = data.frame(COVID = CCA_trt$v[,1], nonCOVID = CCA_ctrl$v[,1]),
                          aes(x = COVID, y = nonCOVID))+
  geom_point(color = "cyan2", alpha=0.5)+
  theme_bw()+
  annotate("text",x= 0.1, y= 0.1, label = paste0("Corr = ", round(cor(CCA_trt$v[,1], CCA_ctrl$v[,1]),4)))+
  labs( x = "Loading of proteins from COVID group analysis", y = "Loading of proteins from non COVID group analysis")+
  theme(panel.grid = element_blank(),
        text = element_text(size = 14),  # Adjust text size
        axis.title = element_text(size = 12),  # Adjust axis title size
        axis.text = element_text(size = 12),
        legend.position = "bottom")

# ggsave("../inst/SavedFigures/COVID_Yloading_ta.pdf", plot = p_Yloading_ta_comp, width = 5, height = 5)

p_Yloading_ta_bar <- ggplot(data = data.frame(COVID = CCA_trt$v[,1], nonCOVID = CCA_ctrl$v[,1], item = 1:ncol(Yt)),
                          aes(x = item))+
 geom_bar(aes(x=item, y = sort(abs(COVID))), fill = "tomato", position = "dodge", stat = "identity")+
 geom_bar(aes(x=item, y = -abs(nonCOVID)[order(abs(COVID))]), position = "dodge", stat = "identity", fill= "steelblue1")+   theme_bw()+
 annotate("text", x= 250, y= 0.1, label = "From COVID group analysis", color = "tomato")+
 annotate("text",x= 250, y= -0.11, label = "From non COVID group analysis", color = "steelblue1")+
 labs( x = "Proteins", y = "Absolute loadings of proteins")+
 theme(panel.grid = element_blank(),
        text = element_text(size = 14),  # Adjust text size
        axis.title = element_text(size = 12),  # Adjust axis title size
        axis.text = element_text(size = 12),
               axis.text.x = element_blank(),
        legend.position = "bottom")

# ggsave("../inst/SavedFigures/COVID_Yloading_ta_bar.pdf", plot = p_Yloading_ta_bar, width = 5, height = 5)


p_Xloading_ta_comp <- ggplot(data = data.frame(COVID = CCA_trt$u[,1], nonCOVID = CCA_ctrl$u[,1]),
                             aes(x=COVID, y=nonCOVID))+
  geom_point(color = "purple", alpha=0.5)+
  theme_bw()+
  annotate("text",x= 0.15, y= 0.05, label = paste0("Corr = ", round(cor(CCA_trt$u[,1], CCA_ctrl$u[,1]),4)))+
  labs(x = "Loading of metabolites from COVID group analysis",
       y = "Loading of metabolites from non COVID group analysis")+
  theme(panel.grid = element_blank(),
        text = element_text(size = 14),  # Adjust text size
        axis.title = element_text(size = 12),  # Adjust axis title size
        axis.text = element_text(size = 12),
        legend.position = "bottom")+
  scale_color_d3()

# ggsave("../inst/SavedFigures/COVID_Xloading_ta.pdf", plot = p_Xloading_ta_comp, width = 5, height = 5)

p_Xloading_ta_bar <- ggplot(data = data.frame(COVID = CCA_trt$u[,1], nonCOVID = CCA_ctrl$u[,1], item = 1:ncol(Xt)),
                          aes(x = item))+
 geom_bar(aes(x=item, y = sort(abs(COVID))), fill = "goldenrod2", position = "dodge", stat = "identity")+
 geom_bar(aes(x=item, y = -abs(nonCOVID)[order(abs(COVID))]), position = "dodge", stat = "identity", fill= "aquamarine")+   theme_bw()+
 labs( x = "Metabolites", y = "Loading of metabolites")+
 annotate("text", x= 55, y= 0.15, label = "From COVID group analysis", color = "goldenrod2")+
 annotate("text",x= 55, y= -0.2, label = "From non COVID group analysis", color = "aquamarine")+
 theme(panel.grid = element_blank(),
        text = element_text(size = 14),  # Adjust text size
        axis.title = element_text(size = 12),  # Adjust axis title size
        axis.text = element_text(size = 12),
        axis.text.x = element_blank(),
        legend.position = "bottom")

# ggsave("../inst/SavedFigures/COVID_Xloading_ta_bar.pdf", plot = p_Xloading_ta_bar, width = 5, height = 5)

p_Yloading_tc_comp <- ggplot(data = data.frame(COVID = CCA_trt$v[,1], Contrastive = CCA_ctst$v[,1]),
                             aes(x=COVID, y=Contrastive))+
  geom_point(color = "plum2", alpha=0.5)+
  theme_bw()+
  annotate("text",x= 0.1, y= 0.06, label = paste0("Corr = ", round(cor(CCA_trt$v[,1], CCA_ctst$v[,1]),4)))+
  labs(x = "Loading of proteins from COVID group analysis",
      y = "Loading of proteins from contrastive analysis")+
  theme(panel.grid = element_blank(),
        text = element_text(size = 14),  # Adjust text size
        axis.title = element_text(size = 12),  # Adjust axis title size
        axis.text = element_text(size = 12),
        legend.position = "bottom")+
  scale_color_d3()
# ggsave("../inst/SavedFigures/COVID_Yloading_tc.pdf", plot = p_Yloading_tc_comp, width = 5, height = 5)

p_Yloading_tc_bar <- ggplot(data = data.frame(COVID = CCA_trt$v[,1], nonCOVID = CCA_ctst$v[,1], item = 1:ncol(Yt)),
                          aes(x = item))+
 geom_bar(aes(x=item, y = sort(abs(COVID))), fill = "tomato", position = "dodge", stat = "identity")+
 geom_bar(aes(x=item, y = -abs(nonCOVID)[order(abs(COVID))]), position = "dodge", stat = "identity", fill= "darkgreen")+   theme_bw()+
 labs( x = "Proteins", y = "Loading of proteins")+
 annotate("text", x= 250, y= 0.1, label = "From COVID group analysis", color = "tomato")+
 annotate("text",x= 250, y= -0.13, label = "From contrastive analysis", color = "darkgreen")+
 theme(panel.grid = element_blank(),
        text = element_text(size = 14),  # Adjust text size
        axis.title = element_text(size = 12),  # Adjust axis title size
        axis.text = element_text(size = 12),
               axis.text.x = element_blank(),
        legend.position = "bottom")
# ggsave("../inst/SavedFigures/COVID_Yloading_tc_bar.pdf", plot = p_Yloading_tc_bar, width = 5, height = 5)

p_moti2 <- ggarrange(p_score_t, p_score_a, p_Yloading_ta_bar, nrow=1, common.legend = TRUE, legend = "bottom", labels = "AUTO", align = "h")
ggsave("../inst/SavedFigures/COVID_moti.pdf", plot = p_moti2, width = 13, height = 5)
```





# Real data application of s3CA on Sorghum multi-omics data
```{r sorghum data s3CA}
##### 0. Data organization ####
load("/Users/yunhuiqi/Library/Mobile Documents/com~apple~CloudDocs/PhD/academic/Research/SmallHarvest/GroupSparseCCA/Drought analysis/SH_Drought_102_341_468_wRep.RData")

ancom1 <- function(dat){
  nonzerosum <- c() 
  nonzerocount <- c()
  for (i in 1:(dim(dat)[1])) {
    sum <- 0
    count <- 0
    for (k in 1:(dim(dat)[2])) {
      if(dat[i,k]==0){
        dat[i,k] <- log(dat[i,k]+1)
      }
      else{
        dat[i,k] <- log(dat[i,k]+1)
        sum=sum+dat[i,k]
        count=count+1
      }
    }
    nonzerosum=c(nonzerosum,sum)
    nonzerocount=c(nonzerocount,count)
  }
  geomean <- matrix(rep.int(nonzerosum/nonzerocount,times = rep(dim(dat)[2],dim(dat)[1])),nrow = dim(dat)[1],byrow = T) 
  return(dat-geomean)
}

# reorder by group information
mtbgroup <- mtb_group[order(mtb_group$Superclass),]
mbgroup <- mb_group[order(mb_group$phylum),]

mtb <- mtb[,mtbgroup$mtb]
mb <- mb[,mbgroup$OTU]

# check align mtb and mb
covariates <- meta
rownames(covariates) <- meta$Barcode
covariates$Date <- factor(covariates$Date, levels = c(70717, 80117, 82817), labels = c("July07","Aug01","Aug28"))

# transformation
mtb_log <- log(mtb + 1)
mb_ancom1 <- ancom1(mb + 1)


#### 1. Choose correlation or covariance ####
# Use correlation matrix or covariance matrix
cormat <- TRUE
#### 2. Choose conditions ####
a_ind <- covariates$Treatment == "WW"
t_ind <- covariates$Treatment == "WS"

X_ind <- 1:ncol(mtb_log)
Y_ind <- 1:ncol(mb_ancom1)

Xa <- scale(mtb_log[a_ind,X_ind], scale = cormat)
Ya <- scale(mb_ancom1[a_ind,Y_ind], scale = cormat)
Xt <- scale(mtb_log[t_ind,X_ind], scale = cormat)
Yt <- scale(mb_ancom1[t_ind,Y_ind], scale = cormat)

p <- ncol(Xa)
q <- ncol(Ya)
na <- nrow(Xa)
nt <- nrow(Xt)
Kmax <- min(p, q)
covariates_a <- covariates[rownames(Xa),]
covariates_t <- covariates[rownames(Xt),]

#### 3. s3CA  ####
res <- cCCA(Xa = Xa, Ya = Ya, Xt = Xt, Yt = Yt, eta = NULL, sparse_ctst = 0.2, sparse_trt = 0.2, sparse_ctrl = 0.2)
cpmd <- res$CCA_ctst
pmd_t <- res$CCA_trt
pmd_a <- res$CCA_ctrl

#### 4. Score plot ####

# contrastive score df 
df_score <- data.frame(X_C1 = Xt%*%cpmd$u[,1],
                       X_C2 = Xt%*%cpmd$u[,2],
                       Y_C1 = Yt%*%cpmd$v[,1],
                       Y_C2 = Yt%*%cpmd$v[,2],
                       covariates_t)

df_score$cluster <- factor(kmeans(df_score[,c("X_C1","X_C2")], centers = 2)$cluster)

# treatment score df
df_score_t <- data.frame(X_C1 =  Xt%*%pmd_t$u[,1], X_C2 =  Xt%*%pmd_t$u[,2],
                         Y_C1 =  Yt%*%pmd_t$v[,1], Y_C2 =  Yt%*%pmd_t$v[,2],
                         covariates_t)

df_score_a <- data.frame(X_C1 =  Xa%*%pmd_a$u[,1], X_C2 =  Xa%*%pmd_a$u[,2],
                         Y_C1 =  Ya%*%pmd_a$v[,1], Y_C2 =  Ya%*%pmd_a$v[,2],
                         covariates_a)


p_score <- ggplot(df_score)+
  geom_point(aes(x = X_C1, y = X_C2, color = factor(cluster),shape = Date), size=2)+
  labs(x = "First component wrt metabolites", y = "Second component wrt metabolites", color = "Contrastive clusters", title = "Contrasitve")+
  theme_bw()+
  theme(panel.grid = element_blank(),
        text = element_text(size = 14),  # Adjust text size
        axis.title = element_text(size = 16),  # Adjust axis title size
        axis.text = element_text(size = 12),
        legend.position = "right")+
  scale_color_aaas()

# ggsave("../inst/SavedFigures/Sorghum_WSWW.pdf", p_score, width = 8, height = 6, unit = "in")


p_score_t_wcs <-  ggplot(df_score_t)+
  geom_point(aes(x = X_C1, y = X_C2, color = factor(df_score$cluster), shape = Date), size=2)+
  labs(x = "First component wrt metabolites", y = "Second component wrt metabolites", color = "Contrastive clusters", title = "Water-stressed only")+
  theme_bw()+
  theme(panel.grid = element_blank(),
        text = element_text(size = 14),  # Adjust text size
        axis.title = element_text(size = 16),  # Adjust axis title size
        axis.text = element_text(size = 12),
        legend.position = "right")+
  scale_color_aaas()

p_score_a_ws <-  ggplot(df_score_a)+
  geom_point(aes(x = X_C1, y = X_C2, shape = Date), size=2)+
  labs(x = "First component wrt metabolites", y = "Second component wrt metabolites", title = "Well-watered only")+
  theme_bw()+
  theme(panel.grid = element_blank(),
        text = element_text(size = 14),  # Adjust text size
        axis.title = element_text(size = 16),  # Adjust axis title size
        axis.text = element_text(size = 12),
        legend.position = "right")+
  scale_color_aaas()

# ggsave("../inst/SavedFigures/Sorghum_WS_wcs.pdf", p_score_t_wcs, width = 8, height = 6, unit = "in")
# ggsave("../inst/SavedFigures/Sorghum_score.pdf", ggarrange(p_score, p_score_t_wcs, p_score_a_ws, common.legend = TRUE, legend = "bottom", nrow = 1), width = 12, height = 5, unit = "in")

p_score_t <-  ggplot(df_score_t)+
  geom_point(aes(x = X_C1, y = X_C2), size=2)+
  labs(x = "First component wrt metabolites", y = "Second component wrt metabolites")+
  theme_bw()+
  theme(panel.grid = element_blank(),
        text = element_text(size = 14),  # Adjust text size
        axis.title = element_text(size = 16),  # Adjust axis title size
        axis.text = element_text(size = 12),
        legend.position = "right")
# ggsave("../inst/SavedFigures/Sorghum_WS.pdf", p_score_t, width = 6, height = 6, unit = "in")

print(ggarrange(p_score, p_score_t_wcs, p_score_a_ws, common.legend = TRUE, legend = "bottom"))

#### 5. Loading plot ####
# loading groups
nb.cols <- 15
mycolors <- colorRampPalette(brewer.pal(20, "Set3"))(nb.cols)
mtb_bar_t <- data.frame(Loadings=c(pmd_t$u[,2]),
                        Superclass=rep(mtbgroup$Superclass,1),
                        Metabolites_Index=1:ncol(Xt),
                        Unique = (pmd_t$u[,2] !=0 & cpmd$u[,2]==0),
                        Analysis = "Water-stressed only")
mtb_bar_t$min=rep(c(0,cumsum(table(mtb_bar_t$Superclass))[1:(nb.cols-1)])+0.5,table(mtb_bar_t$Superclass))
mtb_bar_t$max=rep(c(cumsum(table(mtb_bar_t$Superclass))[1:nb.cols])+0.5,table(mtb_bar_t$Superclass))
mtb_bar_ctst <- data.frame(Loadings=cpmd$u[,2],
                          Superclass=rep(mtbgroup$Superclass,1),
                          Metabolites_Index=1:ncol(Xt),
                          Unique = (pmd_t$u[,2] ==0 & cpmd$u[,2]!=0),
                          Analysis = "Contrastive")
mtb_bar_ctst$min=rep(c(0,cumsum(table(mtb_bar_ctst$Superclass))[1:(nb.cols-1)])+0.5,table(mtb_bar_ctst$Superclass))
mtb_bar_ctst$max=rep(c(cumsum(table(mtb_bar_ctst$Superclass))[1:nb.cols])+0.5,table(mtb_bar_ctst$Superclass))

mtb_bar <- rbind(mtb_bar_t, mtb_bar_ctst)


p_mtbbar <- ggplot(mtb_bar,aes(x = Metabolites_Index , y = Loadings, fill = Unique)) +
  geom_rect(data = mtb_bar,aes(xmin = min, xmax = max, ymin = -Inf, ymax =Inf, fill = Superclass),show.legend = F)+
  theme_minimal()+
  labs(title="")+
  facet_wrap(vars(Analysis), nrow = 2)+
  geom_bar(aes(x = Metabolites_Index , y = Loadings),position = "dodge", stat = "identity", show.legend = T)+
  theme(legend.title = element_blank(), legend.position = "top",legend.text = element_text( size = 8))+
  theme(axis.text.x = element_blank())+
  scale_fill_manual(values = c("TRUE" = "red","FALSE" = "black",
                               "Alkaloids and derivatives"=mycolors[1],
                               "Benzenoids"=mycolors[2],
                               "Carboxylic acids and derivatives"=mycolors[3],
                               "Homogeneous non-metal compounds"=mycolors[4],
                               "Lignans, neolignans and related compounds" =mycolors[5],
                               "Lipids and lipid-like molecules"=mycolors[6],
                               "Nucleosides, nucleotides, and analogues"=mycolors[7],
                               "Organic acids and derivatives" =mycolors[8],
                               "Organic compounds" =mycolors[9],
                               "Organic nitrogen compounds" =mycolors[10],
                               "Organic oxygen compounds" =mycolors[11],
                               "Organoheterocyclic compounds"=mycolors[12],
                               "Organooxygen compounds"=mycolors[13],
                               "Others"=mycolors[14],
                               "Phenylpropanoids and polyketides"=mycolors[15]),
                    breaks = sort(unique(mtb_bar$Superclass)))+
                    #guide = "none")+
  theme(strip.text.x = element_text(size = 10))+
  theme(axis.title =element_text(size=14))+
  theme(plot.title  = element_text(hjust = 0.5))+
  guides(linetype = guide_legend(override.aes = list(fill = c("black","white"))))+
  theme(legend.key = element_rect(fill = "white", colour = "black"))

print(p_mtbbar)
# ggsave("../inst/SavedFigures/Sorghum_loadingsmtb.pdf", p_mtbbar, width = 12, height = 8, unit = "in")


nb.cols <- 17
mycolors <- colorRampPalette(brewer.pal(8, "Set3"))(nb.cols)
mb_bar_t <- data.frame(Loadings=c(pmd_t$v[,2]),
                       Phylum=rep(mbgroup$phylum,1),
                       Microbes_Index=rep(1:468,1),
                       Unique = (pmd_t$v[,2] !=0 & cpmd$v[,2]==0),
                       Analysis = "Water-stressed only")
mb_bar_t$min=rep(c(0,cumsum(table(mb_bar_t$Phylum))[1:(nb.cols-1)])+0.5,table(mb_bar_t$Phylum))
mb_bar_t$max=rep(c(cumsum(table(mb_bar_t$Phylum))[1:nb.cols])+0.5,table(mb_bar_t$Phylum))
mb_bar_ctst <- data.frame(Loadings=cpmd$v[,2],
                          Phylum=rep(mbgroup$phylum,1),
                          Microbes_Index=rep(1:468,1),
                          Unique = (pmd_t$v[,2] ==0 & cpmd$v[,2]!=0),
                          Analysis = "Contrastive")
mb_bar_ctst$min=rep(c(0,cumsum(table(mb_bar_ctst$Phylum))[1:(nb.cols-1)])+0.5,table(mb_bar_ctst$Phylum))
mb_bar_ctst$max=rep(c(cumsum(table(mb_bar_ctst$Phylum))[1:nb.cols])+0.5,table(mb_bar_ctst$Phylum))

mb_bar <- rbind(mb_bar_t, mb_bar_ctst)



p_mbbar <- ggplot(mb_bar,aes(x =Microbes_Index , y = Loadings, fill=Unique)) +
  geom_rect(data = mb_bar,aes(xmin = min, xmax = max, ymin = -Inf, ymax =Inf, fill = Phylum),show.legend = F)+
  theme_minimal()+
  labs(title="")+
  geom_bar(aes(x =Microbes_Index , y = Loadings),position = "dodge", stat = "identity", show.legend = T)+
  theme(legend.title = element_blank(), legend.position = "top",legend.text = element_text( size = 8))+
  theme(axis.text.x = element_blank())+
  facet_wrap(vars(Analysis), nrow = 2)+
  scale_fill_manual(values = c("TRUE" = "red","FALSE" = "black",
                               "Acidobacteria"=mycolors[1],
                               "Actinobacteria"=mycolors[2],
                               "Armatimonadetes"=mycolors[3],
                               "Bacteroidetes"=mycolors[4],
                               "candidate_division_WPS"=mycolors[5],
                               "Chloroflexi" =mycolors[6],
                               "Cyanobacteria"=mycolors[7],
                               "Firmicutes"=mycolors[8],
                               "Gemmatimonadetes"=mycolors[9],
                               "Ignavibacteriae"=mycolors[10],
                               "Nitrospirae"   =mycolors[11],
                               "Planctomycetes"    =mycolors[12],
                               "Poribacteria" =mycolors[13],
                               "Proteobacteria"=mycolors[14],
                               "Tenericutes"=mycolors[15],
                               "Thaumarchaeota"   =mycolors[16],
                               "Verrucomicrobia"    =mycolors[17]),
                    # breaks = sort(unique(mb_bar$Sig_loading)))+
                    breaks = sort(unique(mb_bar$Phylum)))+
                    #guide = "none")+
  theme(strip.text.x = element_text(size = 14))+
  theme(axis.title =element_text(size=14))+
  theme(plot.title  = element_text(hjust = 0.5))+
  theme(legend.key = element_rect(fill = "white", colour = "black"))+
  scale_x_discrete(limits=levels(mb_bar$Microbiome_Index))+
  guides(fill=guide_legend(nrow=3,byrow=TRUE))

p_mbbar

# ggsave("../inst/SavedFigures/Sorghum_loadingsmb.pdf", p_mbbar, width = 12, height = 8, unit = "in")


#### 6. Expression heatmap ####
loading_type <- case_when(cpmd$u[,2]==0 & pmd_t$u[,2]!=0 ~ "Water-stressed",
                          cpmd$u[,2]!=0 & pmd_t$u[,2]==0 ~ "Contrastive",
                          cpmd$u[,2]==0 & pmd_t$u[,2]==0 ~ "none",
                          cpmd$u[,2]!=0 & pmd_t$u[,2]!=0 ~ "both")
heat_dat <- Xt[,loading_type != "none"]
loading_type <- loading_type[loading_type != "none"]
pdf("../inst/SavedFigures/Sorghum_heatmap.pdf", width = 8, height= 6)
p_heat <- Heatmap(t(heat_dat), column_split = ifelse(df_score$cluster == 1, "Contrastive cluster1", "Contrastive cluster2") ,
        cluster_rows = TRUE, cluster_columns = TRUE, row_split = loading_type, 
        show_row_names = FALSE, show_column_names = FALSE, name = "Standardized expression",
        heatmap_legend_param = list(direction = "horizontal"))
draw(p_heat,  heatmap_legend_side = "bottom")
dev.off()

```

